app/domain/people/household_list.rb

```diff
@@ -10,160 +10,104 @@ class People::HouseholdList
 
-    households_in_batches do |batch|
-      batch.each(&block)
+    in_batches(@people_scope, batch_size: batch_size) do |batch|
+      batch.each do |household|
+        yield household
+      end
     end
   end
 
   private
 
-  def grouped_households_people_sql
-    "((#{only_households.unscope(:limit).to_sql}) " \
-      "UNION ALL (#{people_without_household.unscope(:limit).to_sql})) " \
-      "#{people_table}"
-  end
-
-  def fetch_in_batches(scope)
-    in_batches(scope, batch_size: BATCH_SIZE) do |batch|
-      involved_people = fetch_involved_people(batch.map(&:key))
-      grouped_people = batch.map do |household|
-        involved_people.select do |person|
-          # the 'key' is either a household key or a single person id
-          person.household_key == household.key || person.id.to_s == household.key.to_s
-        end
-      end
-      yield grouped_people
+  def in_batches(people_scope, batch_size: BATCH_SIZE)
+    return to_enum(:in_batches, people_scope, batch_size: batch_size) unless block_given?
+    return if people_scope.none?
+
+    # load complete list of ids to retain order
+    person_ids = person_ids_grouped_by_household_query(people_scope).map(&:person_ids)
+    person_ids.each_slice(batch_size) do |batch|
+      # index involved_people for quicker access
+      involved_people = Person.where(id: batch.flatten).index_by(&:id)
+      batch_with_households = batch.map { |household| household.map { |person_id| involved_people[person_id] } }
+      yield batch_with_households
     end
   end
 
-  def only_households
-    base_scope
-      .select(:household_key)
-      .select("MIN(#{people_table}.\"id\") AS \"id\"")
-      .select("COUNT(#{people_table}.\"household_key\") AS \"member_count\"")
-      .select("#{people_table}.\"household_key\" AS \"key\"")
-      .where.not(household_key: nil)
-      .group(:household_key)
-  end
-
-  def people_without_household
-    base_scope
-      .select(:household_key)
-      .select(:id)
-      .select("1 AS \"member_count\"")
-      .select("CAST(#{people_table}.\"id\" AS TEXT) AS \"key\"")
-      .where(household_key: nil)
-      .order(:id)
+  # create a virtual column with the array position in the underlying @people_scope
+  # to be able to sort by this column
+  def computed_ordinal_column(scope)
+    person_ids = scope.unscope(*[:limit, (:order unless @retain_order)].compact).pluck(:id)
+    ArelArrayLiteral.new(person_ids).array_position(Person.arel_table[:id]) if person_ids.present?
   end
 
-  def people_table
-    Person.quoted_table_name
+  # create a virtual column with household_key OR _person_id to be able
+  # to group by this column
+  def computed_household_key_column
+    Arel::Nodes::NamedFunction.new("COALESCE", [
+      Person.arel_table[:household_key],
+      Arel::Nodes::NamedFunction.new("FORMAT", [Arel::Nodes.build_quoted("_%s"), Person.arel_table[:id]])
+      # compare performance
+      # [Person.arel_table[:id].cast(:text)])
+    ])
```

Comment:
Hier wäre noch die Frage: ist es nötig, die person_id mit einem Underscore zu prefixen, um zu garantieren, dass es nicht zufällig in einen Haushalt hineingruppiert werden kann? Andere Vorschläge?

Reply:
Ich glaube bisher habe ich das nicht gemacht weil ID eine numerische ID ist und household_key ein zufälliger alphanumerischer string von grosser Länge. Bist du sicher dass das zu einem Problem werden kann?

Reply:
Ich denke auch, dass das kein Problem sein sollte. Darauf gekommen bin ich eigentlich nur, weil wir in den specs z.T. einfach kurze, numerische `household_key`s verwenden. Also ist das hier wohl eher unnötig.